File: src/main/java/dev/kikugie/elytratrims/common/recipe/DelegatedRecipe.java
package dev.kikugie.elytratrims.common.recipe;

import net.minecraft.inventory.Inventory;
import net.minecraft.item.ItemStack;
import net.minecraft.recipe.SpecialCraftingRecipe;
import net.minecraft.recipe.book.CraftingRecipeCategory;
import net.minecraft.util.Identifier;
import net.minecraft.world.World;

public abstract class DelegatedRecipe extends SpecialCraftingRecipe {
    public DelegatedRecipe(Identifier id, CraftingRecipeCategory category) {
        /*? if <=1.20.4*/
        super(id, category);
        /*? if >1.20.4*/
        /*super(category);*/
    }

    abstract boolean matches(Inventory inventory);
    abstract ItemStack craft(Inventory inventory);

    /*? <=1.20.4 {*/
    @Override
    public boolean matches(net.minecraft.inventory.CraftingInventory inventory, World world) {
        return matches(inventory);
    }

    @Override
    public ItemStack craft(net.minecraft.inventory.CraftingInventory inventory, net.minecraft.registry.DynamicRegistryManager registryManager) {
        return craft(inventory);
    }
    /*?} else {*//*
    @Override
    public boolean matches(net.minecraft.inventory.RecipeInputInventory inventory, World world) {
        return matches(inventory);
    }

    @Override
    public ItemStack craft(net.minecraft.inventory.RecipeInputInventory inventory, net.minecraft.registry.RegistryWrapper.WrapperLookup lookup) {
        return craft(inventory);
    }
    *//*?} */
}

File: src/main/kotlin/dev/kikugie/elytratrims/common/recipe/DelegatedRecipe.kt
package dev.kikugie.elytratrims.common.recipe

import net.minecraft.item.ItemStack
import net.minecraft.recipe.ShapelessRecipe
import net.minecraft.recipe.book.CraftingRecipeCategory
import net.minecraft.util.Identifier
import net.minecraft.util.collection.DefaultedList
import net.minecraft.world.World

abstract class DelegatedRecipe(
    @JvmField val id: Identifier,
    @JvmField val output: ItemStack,
) : ShapelessRecipe(/*? if <1.21 >>*/id, "impossible", CraftingRecipeCategory.EQUIPMENT, output, DefaultedList.of()) {
    abstract fun matches(input: Stacks): Boolean
    abstract fun craft(input: Stacks): ItemStack

    //? if >=1.21 {
    /*override fun craft(input: net.minecraft.recipe.input.CraftingRecipeInput, lookup: net.minecraft.registry.RegistryWrapper.WrapperLookup): ItemStack =
        craft(input.stacks)
    override fun matches(input: net.minecraft.recipe.input.CraftingRecipeInput, world: World): Boolean =
        matches(input.stacks)
    *///?} else {
    override fun craft(inventory: net.minecraft.inventory.RecipeInputInventory, lookup: net.minecraft.registry.DynamicRegistryManager): ItemStack =
        craft(inventory.inputStacks)
    override fun matches(inventory: net.minecraft.inventory.RecipeInputInventory, world: World): Boolean =
        matches(inventory.inputStacks)
    //?}
}